File: src/main/java/io/scriptor/isa/Instruction.java
package io.scriptor.isa;

import org.jetbrains.annotations.NotNull;

import java.util.NoSuchElementException;
import java.util.function.IntPredicate;

public record Instruction(
        @NotNull String mnemonic,
        int ilen,
        int mask,
        int bits,
        int restriction,
        @NotNull Operand[] operands
) implements IntPredicate {

    @Override
    public boolean test(final int value) {
        if (bits != (value & mask)) {
            return false;
        }
        for (final var operand : operands) {
            if (operand.excludes(value)) {
                return false;
            }
        }
        return true;
    }

    public int decode(final int instruction, final @NotNull String label) {
        for (final var operand : operands) {
            if (operand.label().equals(label)) {
                return operand.extract(instruction);
            }
        }
        throw new NoSuchElementException(label);
    }

    public void decode(final int instruction, final int @NotNull [] values, final @NotNull String label0) {
        if (values.length < 1) {
            throw new IllegalArgumentException();
        }

        values[0] = decode(instruction, label0);
    }

    public void decode(
            final int instruction,
            final int @NotNull [] values,
            final @NotNull String label0,
            final @NotNull String label1
    ) {
        if (values.length < 2) {
            throw new IllegalArgumentException();
        }

        values[0] = decode(instruction, label0);
        values[1] = decode(instruction, label1);
    }

    public void decode(
            final int instruction,
            final int @NotNull [] values,
            final @NotNull String label0,
            final @NotNull String label1,
            final @NotNull String label2
    ) {
        if (values.length < 3) {
            throw new IllegalArgumentException();
        }

        values[0] = decode(instruction, label0);
        values[1] = decode(instruction, label1);
        values[2] = decode(instruction, label2);
    }

    @Override
    public @NotNull String toString() {
        return "%s [%d -> %08x & %08x] %s".formatted(mnemonic, ilen, bits, mask, operands);
    }
}

File: src/main/java/io/scriptor/isa/Instruction.kt
package io.scriptor.isa

import io.scriptor.util.Log.format

data class Instruction(
    val mnemonic: String,
    val ilen: UInt,
    val mask: Int,
    val bits: Int,
    val restriction: UInt,
    val operands: Map<String, Operand>,
) {

    fun test(mode: UInt, value: Int): Boolean {
        when {
            restriction != 0U && restriction != mode -> {
                return false
            }

            bits != (value and mask) -> {
                return false
            }

            else -> {
                for (operand in operands.values) {
                    if (operand.excludes(value)) {
                        return false
                    }
                }
                return true
            }
        }
    }

    fun decode(instruction: Int, label: String): Int = when (label) {
        in operands -> operands[label]!!.decode(instruction)
        else -> throw NoSuchElementException(label)
    }

    fun encode(instruction: Int, label: String, value: Int): Int = when (label) {
        in operands -> operands[label]!!.encode(instruction, value)
        else -> throw NoSuchElementException(label)
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun decode(instruction: UInt, values: UIntArray, label0: String) {
        require(values.size >= 1)

        values[0] = decode(instruction.toInt(), label0).toUInt()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun decode(instruction: UInt, values: UIntArray, label0: String, label1: String) {
        require(values.size >= 2)

        values[0] = decode(instruction.toInt(), label0).toUInt()
        values[1] = decode(instruction.toInt(), label1).toUInt()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun decode(instruction: UInt, values: UIntArray, label0: String, label1: String, label2: String) {
        require(values.size >= 3)

        values[0] = decode(instruction.toInt(), label0).toUInt()
        values[1] = decode(instruction.toInt(), label1).toUInt()
        values[2] = decode(instruction.toInt(), label2).toUInt()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun decode(instruction: UInt, values: UIntArray, label0: String, label1: String, label2: String, label3: String) {
        require(values.size >= 4)

        values[0] = decode(instruction.toInt(), label0).toUInt()
        values[1] = decode(instruction.toInt(), label1).toUInt()
        values[2] = decode(instruction.toInt(), label2).toUInt()
        values[3] = decode(instruction.toInt(), label3).toUInt()
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun decode(
        instruction: UInt,
        values: UIntArray,
        label0: String,
        label1: String,
        label2: String,
        label3: String,
        label4: String,
    ) {
        require(values.size >= 5)

        values[0] = decode(instruction.toInt(), label0).toUInt()
        values[1] = decode(instruction.toInt(), label1).toUInt()
        values[2] = decode(instruction.toInt(), label2).toUInt()
        values[3] = decode(instruction.toInt(), label3).toUInt()
        values[4] = decode(instruction.toInt(), label4).toUInt()
    }

    override fun toString(): String {
        return format(
            "%s [%d -> %08x & %08x] %s",
            mnemonic,
            ilen,
            bits,
            mask,
            operands,
        )
    }
}
